pkg/metrics/stats_reporter.go

```diff
@@ -0,0 +1,163 @@
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"context"
+	"strconv"
+	"sync"
+	"time"
+
+	"go.opentelemetry.io/otel/api/global"
+	"go.opentelemetry.io/otel/api/metric"
+	"go.opentelemetry.io/otel/label"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/secrets-store-csi-driver/apis/v1alpha1"
+)
+
+var (
+	secretProviderClassKey    = "secretproviderclass"
+	namespaceKey              = "namespace"
+	secretObjectName          = "secret_name"
+	secretObjectTypeKeyPrefix = "secret_type_"
+	providerKey               = "provider"
+	labelKeyPrefix            = "label_"
+
+	metricInfoName                    = "kube_secretproviderclass_info"
+	metricInfoDesc                    = "Information about SecretProviderClass"
+	metricTypeName                    = "kube_secretproviderclass_type"
+	metricTypeDesc                    = "Type about SecretProviderClass"
+	metricLabelsName                  = "kube_secretproviderclass_labels"
+	metricLabelsDesc                  = "Kubernetes labels converted to OpenTelemetry labels"
+	metricCreatedName                 = "kube_secretproviderclass_created"
+	metricCreatedDesc                 = "Unix creation timestamp"
+	metricMetadataResourceVersionName = "kube_secretproviderclass_metadata_resource_version"
+	metricMetadataResourceVersionDesc = "Resource version representing a specific version of secret"
+)
+
+type reporter struct {
+	client.Client
+}
+
+func NewSecretProviderClassReporter(client client.Client) manager.Runnable {
+	return &reporter{
+		Client: client,
+	}
+}
+
+func (r *reporter) Start(ctx context.Context) error {
+	ticker := time.NewTicker(5 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-ticker.C:
+			if err := r.report(ctx); err != nil {
+				klog.ErrorS(err, "failed to scrape secretproviderclass metrics")
+			}
+		}
+	}
+}
+
+func (r *reporter) report(ctx context.Context) error {
+	meter := global.Meter("secretsstore")
+	var (
+		err                           error
+		metricInfo                    metric.Int64ValueObserver
+		metricType                    metric.Int64ValueObserver
+		metricLabel                   metric.Int64ValueObserver
+		metricCreated                 metric.Float64ValueObserver
+		metricMetadataResourceVersion metric.Float64ValueObserver
+		// lock prevents a race between batch observer and instrument registration.
+		lock sync.Mutex
```

Comment:
arn't these scoped to only the individual `report()` calls? Should they be members of the `reporter` struct instead?

Reply:
moved them to struct.